File: frontend/src/types/schedule.ts
export enum Shift {
  MORNING = 'morning',
  AFTERNOON = 'afternoon',
  EVENING = 'evening'
}

export enum AttendanceStatus {
  PENDING = 'pending',
  PRESENT = 'present',
  ABSENT = 'absent',
  LATE = 'late',
  EARLY_LEAVE = 'early_leave',
  EXTRA = 'extra'
}

export interface Schedule {
  id: string;
  user_id: string;
  work_date: string; // YYYY-MM-DD
  shift: Shift;
  registration_type: 'manual' | 'auto';
  is_cancelled: boolean;
  cancelled_at?: string;
  cancel_reason?: string;
  created_at: string;
  updated_at: string;
}

export interface ClassSchedule {
  id: string;
  subject_name: string;
  room?: string;
  start_datetime: string;
  end_datetime: string;
  is_cancelled: boolean;
  description?: string;
}


export interface ScheduleCreate {
  work_date: string;
  shift: Shift;
}

export interface Attendance {
  id: string;
  user_id: string;
  schedule_id?: string;
  work_date: string;
  shift: Shift;
  check_in_time?: string;
  check_out_time?: string;
  status: AttendanceStatus;
  discipline_points_change: number;
  bonus_points: number;
  auto_reconciled: boolean;
  reconciled_at?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface AttendanceCheckIn {
  shift: Shift;
  notes?: string;
}

export interface AttendanceCheckOut {
  notes?: string;
}

export interface AttendanceStats {
  total_scheduled: number;
  total_attended: number;
  total_absent: number;
  total_late: number;
  total_extra: number;
  attendance_rate: number;
  discipline_points_total: number;
  bonus_points_total: number;
}

export interface WeekScheduleRequest {
  schedules: ScheduleCreate[];
}

export interface WeekScheduleResponse {
  week_start: string;
  week_end: string;
  schedules: Schedule[];
  class_schedules: ClassSchedule[];
  attendances: Attendance[];
  stats: AttendanceStats | null;
}

export interface ForceSyncFailedUser {
  user_id: string;
  email: string;
  student_id: string;
  error: string;
}

export interface ForceSyncAllUsersResponse {
  success: boolean;
  message: string;
  week_start?: string;
  total_users?: number;
  success_count?: number;
  error_count?: number;
  failed_users?: ForceSyncFailedUser[];
  error?: string;
}

export const SHIFT_LABELS: Record<Shift, string> = {
  [Shift.MORNING]: 'Sáng (8h-12h)',
  [Shift.AFTERNOON]: 'Chiều (13h-17h)',
  [Shift.EVENING]: 'Tối (18h-22h)'
};

export const ATTENDANCE_STATUS_LABELS: Record<AttendanceStatus, string> = {
  [AttendanceStatus.PENDING]: 'Chờ xác nhận',
  [AttendanceStatus.PRESENT]: 'Có mặt',
  [AttendanceStatus.ABSENT]: 'Vắng mặt',
  [AttendanceStatus.LATE]: 'Đi muộn',
  [AttendanceStatus.EARLY_LEAVE]: 'Về sớm',
  [AttendanceStatus.EXTRA]: 'Làm thêm'
};

export const ATTENDANCE_STATUS_COLORS: Record<AttendanceStatus, string> = {
  [AttendanceStatus.PENDING]: 'bg-gray-100 text-gray-800',
  [AttendanceStatus.PRESENT]: 'bg-green-100 text-green-800',
  [AttendanceStatus.ABSENT]: 'bg-red-100 text-red-800',
  [AttendanceStatus.LATE]: 'bg-yellow-100 text-yellow-800',
  [AttendanceStatus.EARLY_LEAVE]: 'bg-orange-100 text-orange-800',
  [AttendanceStatus.EXTRA]: 'bg-purple-100 text-purple-800'
};

File: backend/app/schemas/schedule.py
"""
Schedule Schemas - Request/Response models
"""
from datetime import date, datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from uuid import UUID
from app.models.schedule import Shift, AttendanceStatus


# ============================================
# Schedule Schemas
# ============================================

class ScheduleCreate(BaseModel):
    """Schema for creating a schedule registration"""
    work_date: date
    shift: Shift
    
    class Config:
        use_enum_values = True


class ScheduleUpdate(BaseModel):
    """Schema for updating a schedule"""
    is_cancelled: bool = True
    cancel_reason: Optional[str] = None


class ScheduleResponse(BaseModel):
    """Schedule information"""
    id: UUID
    user_id: UUID
    work_date: date
    shift: Shift
    registration_type: str = "manual"
    registered_at: datetime
    is_cancelled: bool
    cancelled_at: Optional[datetime] = None
    cancel_reason: Optional[str] = None
    created_at: datetime
    
    class Config:
        from_attributes = True
        use_enum_values = True


class ClassScheduleResponse(BaseModel):
    """Class schedule information"""
    id: UUID
    subject_name: str
    room: Optional[str] = None
    start_datetime: datetime
    end_datetime: datetime
    is_cancelled: bool
    description: Optional[str] = None
    
    class Config:
        from_attributes = True


class ScheduleListResponse(BaseModel):
    """Paginated schedule list"""
    schedules: List[ScheduleResponse]
    total: int


class WeekScheduleRequest(BaseModel):
    """Batch create schedules for a week"""
    schedules: List[ScheduleCreate]


# ============================================
# Attendance Schemas
# ============================================

class AttendanceCheckIn(BaseModel):
    """Check-in request"""
    shift: Shift
    notes: Optional[str] = None
    
    class Config:
        use_enum_values = True


class AttendanceCheckOut(BaseModel):
    """Check-out request"""
    notes: Optional[str] = None


class AttendanceResponse(BaseModel):
    """Attendance information"""
    id: UUID
    user_id: UUID
    schedule_id: Optional[UUID] = None
    work_date: date
    shift: Optional[Shift] = None
    check_in_time: Optional[datetime] = None
    check_out_time: Optional[datetime] = None
    status: AttendanceStatus
    discipline_points_change: int
    bonus_points: int
    notes: Optional[str] = None
    auto_reconciled: bool
    reconciled_at: Optional[datetime] = None
    created_at: datetime
    
    class Config:
        from_attributes = True
        use_enum_values = True


class AttendanceListResponse(BaseModel):
    """Paginated attendance list"""
    attendances: List[AttendanceResponse]
    total: int


class AttendanceStatsResponse(BaseModel):
    """Attendance statistics"""
    total_scheduled: int
    total_attended: int
    total_absent: int
    total_late: int
    total_extra: int
    attendance_rate: float  # Percentage
    discipline_points_total: int
    bonus_points_total: int


class WeekScheduleResponse(BaseModel):
    """Week overview"""
    week_start: date
    week_end: date
    schedules: List[ScheduleResponse]
    class_schedules: List[ClassScheduleResponse] = []
    attendances: List[AttendanceResponse] = []
    stats: Optional[AttendanceStatsResponse] = None
